pkg/nsx/services/securitypolicy/firewall.go

```diff
@@ -353,52 +381,97 @@ func (service *SecurityPolicyService) createOrUpdateSecurityPolicy(obj *v1alpha1
-		staleRules[i].MarkedForDelete = &MarkedForDelete // InfraClient need this field to delete the group
+		staleRules[i].MarkedForDelete = &MarkedForDelete // nsx clients need this field to delete the rules
 	}
 	finalRules = append(finalRules, staleRules...)
 	finalRules = append(finalRules, changedRules...)
 	finalSecurityPolicy.Rules = finalRules
 
 	finalGroups := make([]model.Group, 0)
 	for i := len(staleGroups) - 1; i >= 0; i-- { // Don't use range, it would copy the element
-		staleGroups[i].MarkedForDelete = &MarkedForDelete // InfraClient need this field to delete the group
+		staleGroups[i].MarkedForDelete = &MarkedForDelete // nsx clients need this field to delete the group
 	}
 	finalGroups = append(finalGroups, staleGroups...)
 	finalGroups = append(finalGroups, changedGroups...)
 
 	// WrapHighLevelSecurityPolicy will modify the input security policy, so we need to make a copy for the following store update.
 	finalSecurityPolicyCopy := *finalSecurityPolicy
-	finalSecurityPolicyCopy.Rules = finalSecurityPolicy.Rules
+	finalSecurityPolicyCopy.Rules = finalRules
 
 	if isVpcEnabled(service) {
 		vpcInfo, err := service.getVpcInfo(obj.ObjectMeta.Namespace)
 		if err != nil {
 			return err
 		}
 
-		var finalProjectGroups *[]model.Group
-		var finalProjectShares *[]model.Share
+		finalProjectGroups := make([]model.Group, 0)
+		finalProjectShares := make([]model.Share, 0)
+		projectGroups := make([]model.Group, 0)
+		projectShares := make([]model.Share, 0)
+		for i := len(*nsxProjectShares) - 1; i >= 0; i-- {
+			projectGroups = append(projectGroups, *((*nsxProjectShares)[i].shareGroup))
+			projectShares = append(projectShares, *((*nsxProjectShares)[i].share))
+		}
 
-		if len(*nsxProjectShares) != 0 {
-			projectGroups := make([]model.Group, 0)
-			projectShares := make([]model.Share, 0)
-			// create/update nsx project shares and nsx project level groups
-			for i := len(*nsxProjectShares) - 1; i >= 0; i-- {
-				projectGroups = append(projectGroups, *((*nsxProjectShares)[i].shareGroup))
-				projectShares = append(projectShares, *((*nsxProjectShares)[i].share))
+		// Create/Update nsx project shares and nsx project level groups
+		existingProjectGroups := projectGroupStore.GetByIndex(common.TagScopeSecurityPolicyCRUID, string(obj.UID))
+		changed, stale := common.CompareResources(GroupsPtrToComparable(existingProjectGroups), GroupsToComparable(projectGroups))
+		changedProjectGroups, staleProjectGroups := ComparableToGroups(changed), ComparableToGroups(stale)
+		if len(changedProjectGroups) == 0 && len(staleProjectGroups) == 0 {
+			log.Info("project groups are not changed, skip updating them")
+		}
+
+		// 1.Remove stale nsxRules
+		if len(staleProjectGroups) != 0 {
+			// Before deleting the stale project groups and project shares, we must delete the nsxRules to which the project shares are referencing
+			// The deleted stale nsxRules might also include other stale rules to which the project shares are not referencing,
```

Comment:
You mean, we will delete ALL stale rules in this step, no matter whether it's linked to any project share?

Reply:
yes, it is. I think the final result is to delete all stale rules, so, it doesn't matter we delete them in the beginning or in the middle or at last.  But for the stale rules specific for group share, we must delete them before deleting project share and project group.

Reply:
Of course, we can also identify the exact stale rules for deleted project share, if that, it needs more computation, we need use buildRuleID tag in the stale project group to compare each stale rule list to selected the final stale project rules.

Reply:
we have merged rules in VPC level and project level shares,groups together into one H-API call, so, there is no need to use this intermediate step any more.